bpf/lib/nat.h

```diff
@@ -0,0 +1,866 @@
+struct bpf_elf_map __section_maps SNAT_MAPPING_IPV4 = {
+	.type		= BPF_MAP_TYPE_HASH,
+	.size_key	= sizeof(struct ipv4_ct_tuple),
+	.size_value	= sizeof(struct ipv4_nat_entry),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= SNAT_MAPPING_IPV4_SIZE,
+};
+
+static __always_inline
+struct ipv4_nat_entry *snat_v4_lookup(struct ipv4_ct_tuple *tuple)
+{
+	return __snat_lookup(&SNAT_MAPPING_IPV4, tuple);
+}
+
+static __always_inline int snat_v4_update(struct ipv4_ct_tuple *otuple,
+					  struct ipv4_nat_entry *ostate,
+					  struct ipv4_ct_tuple *rtuple,
+					  struct ipv4_nat_entry *rstate)
+{
+	return __snat_update(&SNAT_MAPPING_IPV4, otuple, ostate,
+			     rtuple, rstate);
+}
+
+static __always_inline void snat_v4_delete(struct ipv4_ct_tuple *otuple,
+					   struct ipv4_ct_tuple *rtuple)
+{
+	__snat_delete(&SNAT_MAPPING_IPV4, otuple, rtuple);
+}
+
+static __always_inline void snat_v4_swap_tuple(struct ipv4_ct_tuple *otuple,
+					       struct ipv4_ct_tuple *rtuple)
+{
+	__builtin_memset(rtuple, 0, sizeof(*rtuple));
+	rtuple->nexthdr = otuple->nexthdr;
+	rtuple->daddr = otuple->saddr;
+	rtuple->saddr = otuple->daddr;
+	rtuple->dport = otuple->sport;
+	rtuple->sport = otuple->dport;
+	rtuple->flags = otuple->flags == NAT_DIR_EGRESS ?
+			NAT_DIR_INGRESS : NAT_DIR_EGRESS;
+}
+
+static __always_inline int snat_v4_reverse_tuple(struct ipv4_ct_tuple *otuple,
+						 struct ipv4_ct_tuple *rtuple)
+{
+	struct ipv4_nat_entry *ostate;
+
+	ostate = snat_v4_lookup(otuple);
+	if (ostate) {
+		snat_v4_swap_tuple(otuple, rtuple);
+		rtuple->daddr = ostate->to_saddr;
+		rtuple->dport = ostate->to_sport;
+	}
+
+	return ostate ? 0 : -1;
+}
+
+static __always_inline void snat_v4_ct_canonicalize(struct ipv4_ct_tuple *otuple)
+{
+	__be32 addr = otuple->saddr;
+
+	otuple->flags = NAT_DIR_EGRESS;
+	/* Workaround #5848. */
+	otuple->saddr = otuple->daddr;
+	otuple->daddr = addr;
+}
+
+static __always_inline void snat_v4_delete_tuples(struct ipv4_ct_tuple *otuple)
+{
+	struct ipv4_ct_tuple rtuple;
+
+	if (otuple->flags & TUPLE_F_IN)
+		return;
+	snat_v4_ct_canonicalize(otuple);
+	if (!snat_v4_reverse_tuple(otuple, &rtuple))
+		snat_v4_delete(otuple, &rtuple);
+}
+
+static __always_inline int snat_v4_new_mapping(struct __sk_buff *skb,
+					       struct ipv4_ct_tuple *otuple,
+					       struct ipv4_nat_entry *ostate)
+{
+	struct ipv4_nat_entry rstate;
+	struct ipv4_ct_tuple rtuple;
+	int ret, retries;
+
+	__builtin_memset(&rstate, 0, sizeof(rstate));
+	__builtin_memset(ostate, 0, sizeof(*ostate));
+
+	rstate.to_daddr = otuple->saddr;
+	rstate.to_dport = otuple->sport;
+
+	ostate->to_saddr = SNAT_IPV4_EXTERNAL;
+	ostate->to_sport = otuple->sport;
+
+	snat_v4_swap_tuple(otuple, &rtuple);
+	rtuple.daddr = SNAT_IPV4_EXTERNAL;
+
+	if (otuple->saddr == SNAT_IPV4_EXTERNAL) {
+		ostate->common.host_local = 1;
+		rstate.common.host_local = ostate->common.host_local;
+	}
+
+#pragma unroll
+	for (retries = 0; retries < 16; retries++) {
```

Comment:
Define so we can change `16` via daemon flag?

Reply:
Makes sense, I'll integrate this into daemon to emit a define, so we can tweak it from there when we see need.